runelite-client/src/main/java/net/runelite/client/game/chatbox/ChatboxTextInput.java

```diff
@@ -0,0 +1,663 @@
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package net.runelite.client.game.chatbox;
+
+import com.google.inject.Inject;
+import java.awt.Rectangle;
+import java.awt.Toolkit;
+import java.awt.datatransfer.DataFlavor;
+import java.awt.datatransfer.StringSelection;
+import java.awt.datatransfer.UnsupportedFlavorException;
+import java.awt.event.KeyEvent;
+import java.awt.event.MouseEvent;
+import java.io.IOException;
+import java.util.function.Consumer;
+import java.util.function.IntPredicate;
+import java.util.function.Predicate;
+import java.util.function.ToIntFunction;
+import javax.swing.SwingUtilities;
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+import net.runelite.api.FontTypeFace;
+import net.runelite.api.FontID;
+import net.runelite.api.WidgetType;
+import net.runelite.api.widgets.JavaScriptCallback;
+import net.runelite.api.widgets.Widget;
+import net.runelite.api.widgets.WidgetPositionMode;
+import net.runelite.api.widgets.WidgetSizeMode;
+import net.runelite.api.widgets.WidgetTextAlignment;
+import net.runelite.client.callback.ClientThread;
+import net.runelite.client.input.KeyListener;
+import net.runelite.client.input.MouseListener;
+
+@Slf4j
+public class ChatboxTextInput extends ChatboxInput implements KeyListener, MouseListener
+{
+	private static final int CURSOR_FLASH_RATE_MILLIS = 1000;
+
+	private final ChatboxPanelManager chatboxPanelManager;
+	private final ClientThread clientThread;
+
+	private static IntPredicate getDefaultCharValidator()
+	{
+		return i -> i >= 32 && i < 127;
+	}
+
+	@Getter
+	private String prompt;
+
+	private StringBuffer value = new StringBuffer();
+
+	@Getter
+	private int cursor = 0;
+
+	@Getter
+	private int cursorEnd = 0;
+
+	private int selectionStart = -1;
+	private int selectionEnd = -1;
+
+	@Getter
+	private IntPredicate charValidator = getDefaultCharValidator();
+
+	@Getter
+	private Runnable onClose = null;
+
+	@Getter
+	private Consumer<String> onDone = null;
+
+	@Getter
+	private Consumer<String> onChanged = null;
+
+	@Getter
+	private int fontID = FontID.QUILL_8;
+
+	// This is a lambda so I can have atomic updates for it's captures
+	private ToIntFunction<MouseEvent> getCharOffset = null;
+	private Predicate<MouseEvent> isInBounds = null;
+
+	private boolean built = false;
+
+	@Inject
+	protected ChatboxTextInput(ChatboxPanelManager chatboxPanelManager, ClientThread clientThread)
+	{
+		this.chatboxPanelManager = chatboxPanelManager;
+		this.clientThread = clientThread;
+	}
+
+	public ChatboxTextInput prompt(String prompt)
+	{
+		this.prompt = prompt;
+		if (built)
+		{
+			clientThread.invoke(this::update);
+		}
+		return this;
+	}
+
+	public ChatboxTextInput value(String value)
+	{
+		this.value = new StringBuffer(value);
+		if (built)
+		{
+			clientThread.invoke(this::update);
+		}
+		return this;
+	}
+
+	public ChatboxTextInput cursorAt(int index)
+	{
+		return cursorAt(index, index);
+	}
+
+	public ChatboxTextInput cursorAt(int s, int e)
```

Comment:
Could we get real variable names here? Even after looking over the code I'm not 100% certain what on earth is going on here, which would likely be massively remedied by informative names.